source/agora/common/TransactionManager.d

```diff
@@ -0,0 +1,283 @@
+/*******************************************************************************
+
+    Contains class for transaction verification
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.TransactionManager;
+import agora.common.Data;
+import agora.common.crypto.Key;
+import agora.common.Transaction;
+import std.digest.sha;
+import vibe.data.json : serializeToJsonString;
+
+/*******************************************************************************
+
+Class for the transactions verification
+
+*******************************************************************************/
+
+public class TransactionManager
+{
+    /// Store of previously processed transactions
+    private Transaction[Hash] txs;
+
+    /// null value of transaction hash
+    public static Hash nullTxHash;
+
+    /// null value of signature
+    public static Signature nullSignature;
+
+    /*******************************************************************************
+
+        Save a `Transaction`
+
+        Params:
+            hash = `Hash` of `Transaction`
+            tx = Type of `Transaction`
+
+    *******************************************************************************/
+
+    public void saveTransaction (Hash hash, Transaction tx)
+    {
+        this.txs[hash] = tx;
+    }
+
+    /*******************************************************************************
+
+        Return hash of transaction
+
+        Params:
+            tx = Type of `Transaction`
+
+        Return:
+            `Hash` of `Transaction`
+
+    *******************************************************************************/
+
+    public Hash getTxHash (Transaction tx)
+    {
+        Hash hash = sha256Of(serializeToJsonString(tx));
```

Comment:
We should not serialize to JSON here, the struct should be directly hashed. See the hashing primitive PR for an example on how to do this.